src/ic-cdk/src/futures.rs

```diff
@@ -1,121 +1,135 @@
+    // the heap management for us. If CLEANUP is set, then we're recovering from
+    // a callback trap, and want to drop the future without executing any more of it;
+    // if previous_trap is set, then we already recovered from a callback trap in a
+    // different callback, and should immediately trap again in this one.
+    //
     /// # Safety
+    ///
     /// This function should only be called by a [Waker] created by [`waker`].
     unsafe fn wake(ptr: *const ()) {
-        // SAFETY: The function contract guarantees that the outer pointer is a Box of the FuturePtr.
-        let boxed_future_ptr_ptr = unsafe { Box::from_raw(ptr as *mut FuturePtr) };
-        let future_ptr: FuturePtr = *boxed_future_ptr_ptr;
-        // SAFETY: The function contract guarantees that the inner pointer is a FuturePtr, and unique.
-        let mut boxed_future = unsafe { Box::from_raw(future_ptr) };
-        // SAFETY: Boxes satisfy the pinning contract and are sound to use in new_unchecked.
-        // The box is never moved out of, but dropped in place.
-        let mut pinned_future = unsafe { Pin::new_unchecked(&mut *boxed_future) };
-        if !super::CLEANUP.load(Ordering::Relaxed)
-            && pinned_future
-                .as_mut()
-                // SAFETY: The pointer attached via `waker` satisfies its own contract.
-                .poll(&mut Context::from_waker(&unsafe { waker::waker(ptr) }))
-                .is_pending()
-        {
-            Box::into_raw(boxed_future_ptr_ptr);
-            Box::into_raw(boxed_future);
+        // SAFETY: The function's contract guarantees that the pointer is an Rc to a WakerState, and that this call takes ownership of the data.
+        let state = unsafe { Rc::from_raw(ptr as *const WakerState) };
+        // Must check CLEANUP *before* previous_trap, as we may be recovering from the following immediate trap.
+        if super::CLEANUP.load(Ordering::Relaxed) {
+            state.previous_trap.set(true);
+        } else if state.previous_trap.get() {
+            crate::trap("Call already trapped");
+        } else {
+            let waker = waker(Rc::clone(&state));
+            let Ok(mut borrow) = state.future.try_borrow_mut() else {
+                // If this is already borrowed, then wake was called from inside poll. There's not a lot we can do about this - we are not
+                // a true scheduler and so cannot immediately schedule another poll, nor can we reentrantly lock the future. So we cancel it.
+                // Don't wake from poll. There's really no reason to do that at all. 
+                return;
+            };
+            let pinned_future = borrow.as_mut();
+            let _ = pinned_future.poll(&mut Context::from_waker(&waker));
         }
     }
 
-    fn wake_by_ref(_: *const ()) {}
-
-    fn drop(_: *const ()) {}
+    /// # Safety
+    ///
+    /// This function should only be called by a [Waker] created by [waker].
+    unsafe fn wake_by_ref(ptr: *const ()) {
```

Comment:
Why is it necessary to also implement wake_by_ref now?

Reply:
Because it's expected to exist and implementations of Future might interact with it; it's not unsafe correctness but just regular correctness.